control-plane/agents/core/src/pool/service.rs

```diff
@@ -1,22 +1,171 @@
 use crate::core::{registry::Registry, specs::ResourceSpecsLocked, wrapper::GetterOps};
 use common::errors::{PoolNotFound, ReplicaNotFound, SvcError};
 use common_lib::{
-    mbus_api::message_bus::v0::{Pools, Replicas},
+    mbus_api::{
+        message_bus::v0::{Pools, Replicas},
+        ReplyError,
+    },
     types::v0::{
         message_bus::{
             CreatePool, CreateReplica, DestroyPool, DestroyReplica, Filter, GetPools, GetReplicas,
-            NodeId, Pool, PoolId, Replica, ShareReplica, UnshareReplica,
+            NexusId, NodeId, Pool, PoolId, Protocol, Replica, ReplicaId, ReplicaOwners,
+            ReplicaShareProtocol, ShareReplica, UnshareReplica, VolumeId,
         },
         store::OperationMode,
     },
 };
+use grpc::{
+    grpc_opts::Context,
+    pool::traits::{CreatePoolInfo, DestroyPoolInfo, PoolOperations},
+    replica::traits::{
+        CreateReplicaInfo, DestroyReplicaInfo, ReplicaOperations, ShareReplicaInfo,
+        UnshareReplicaInfo,
+    },
+};
 use snafu::OptionExt;
+use std::{convert::TryFrom, str::FromStr};
 
 #[derive(Debug, Clone)]
 pub(super) struct Service {
     registry: Registry,
 }
 
+#[tonic::async_trait]
+impl PoolOperations for Service {
+    async fn create(
+        &self,
+        pool: &(dyn CreatePoolInfo + Sync + Send),
+        _ctx: Option<Context>,
+    ) -> Result<Pool, ReplyError> {
+        let req = CreatePool {
+            node: pool.node_id().into(),
+            id: pool.pool_id().into(),
+            disks: match pool.disks() {
+                None => vec![],
+                Some(disks) => disks.iter().map(|i| i.into()).collect(),
+            },
+            labels: pool.labels(),
+        };
+        let pool = self.create_pool(&req).await?;
+        Ok(pool)
+    }
+
+    async fn destroy(
+        &self,
+        pool: &(dyn DestroyPoolInfo + Sync + Send),
+        _ctx: Option<Context>,
+    ) -> Result<(), ReplyError> {
+        let req = DestroyPool {
+            node: pool.node_id().into(),
+            id: pool.pool_id().into(),
+        };
+        self.destroy_pool(&req).await?;
+        Ok(())
+    }
+
+    async fn get(&self, filter: Filter, _ctx: Option<Context>) -> Result<Pools, ReplyError> {
+        let req = GetPools { filter };
+        let pools = self.get_pools(&req).await?;
+        Ok(pools)
+    }
+}
+
+#[tonic::async_trait]
+impl ReplicaOperations for Service {
+    async fn create(
+        &self,
+        req: &(dyn CreateReplicaInfo + Sync + Send),
+        _ctx: Option<Context>,
+    ) -> Result<Replica, ReplyError> {
+        let create_replica = CreateReplica {
```

Comment:
please remove all the unwrap's where possible, non-test code should not have them as much as possible.

Reply:
Done. Removed all possible unwraps.